### PR TITLE
Extended mustach language to support direct array/collection access

### PR DESCRIPTION
**Pull request by uboness**
Customzied our internal `MustacheFactory` to supports array/collection access syntax in the form of `array_or_collection.X` where `array_or_collection` refers to an array or a collection and `X` is the index into it. For example, referring to the "key" field in the 3rd item in the "array" array can be expressed as `array.3.key`.

Note, although not part of the official mustache specs, `mustache.js` support this and users may expect to have this support.

### CODE REVIEW

File: src/main/java/org/elasticsearch/script/mustache/InternalMustacheFactory.java [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+    @Override
+    public void encode(String value, Writer writer) {
+        try {
+            escape(value, writer);
+        } catch (IOException e) {
+            throw new MustacheException("Failed to encode value: " + value);
+        }
+    }
+
+    public static Writer escape(String value, Writer writer) throws IOException {
```
Comment (MaineC) [inline src/main/java/org/elasticsearch/script/mustache/InternalMustacheFactory.java]: While we are re-touching this code anyway - any reason to not re-use either the Jackson JsonStringEncoder or StringEscapeUtils that are on the class path anyway?

File: src/test/java/org/elasticsearch/script/mustache/MustacheScriptEngineTest.java [hunk 2]
```diff
@@ -98,14 +103,60 @@ public void testEscapeJson() throws IOException {
...
+                ImmutableList.of(ImmutableMap.<String, Object>of("key", "foo"), ImmutableMap.<String, Object>of("key", "bar")),
+                ImmutableSet.of(ImmutableMap.<String, Object>of("key", "foo"), ImmutableMap.<String, Object>of("key", "bar")));
+        vars.put("data", data);
+        Object output = engine.execute(engine.compile("{{data.0.key}} {{data.1.key}}"), vars);
+        assertThat(output, notNullValue());
+        assertThat(output, instanceOf(BytesReference.class));
+        BytesReference bytes = (BytesReference) output;
+        assertThat(bytes.toUtf8(), equalTo("foo bar"));
+    }
+
```
Comment (MaineC) [inline src/test/java/org/elasticsearch/script/mustache/MustacheScriptEngineTest.java]: Can we add at least one more test here that shows how this feature can be used in the context of template ES queries?

File: src/test/java/org/elasticsearch/script/mustache/MustacheScriptEngineTest.java [hunk 3]
```diff
@@ -98,14 +103,60 @@ public void testEscapeJson() throws IOException {
                 escaped.append(c);
             }
             StringWriter target = new StringWriter();
-            assertThat(escaped.toString(), equalTo(JsonEscapingMustacheFactory.escape(writer.toString(), target).toString()));
+            assertThat(escaped.toString(), equalTo(InternalMustacheFactory.escape(writer.toString(), target).toString()));
         }
     }
 
+    @Test @Repeat(iterations = 5)
```
Comment (MaineC) [inline src/test/java/org/elasticsearch/script/mustache/MustacheScriptEngineTest.java]: Instead of running the test for five iterations randomly why not deterministically iterate over the three options you want to test below?

File: src/test/java/org/elasticsearch/script/mustache/MustacheScriptEngineTest.java [hunk 4]
```diff
@@ -98,14 +103,60 @@ public void testEscapeJson() throws IOException {
...
+        vars.put("data", data);
+        Object output = engine.execute(engine.compile("{{data.0}} {{data.1}}"), vars);
+        assertThat(output, notNullValue());
+        assertThat(output, instanceOf(BytesReference.class));
+        BytesReference bytes = (BytesReference) output;
+        assertThat(bytes.toUtf8(), equalTo("foo bar"));
+    }
+
+    @Test @Repeat(iterations = 5)
+    public void testArrayInArrayAccess() throws Exception {
```
Comment (MaineC) [inline src/test/java/org/elasticsearch/script/mustache/MustacheScriptEngineTest.java]: See above

File: src/test/java/org/elasticsearch/script/mustache/MustacheScriptEngineTest.java [hunk 5]
```diff
@@ -98,14 +103,60 @@ public void testEscapeJson() throws IOException {
...
+        );
+        vars.put("data", data);
+        Object output = engine.execute(engine.compile("{{data.0.0}} {{data.0.1}}"), vars);
+        assertThat(output, notNullValue());
+        assertThat(output, instanceOf(BytesReference.class));
+        BytesReference bytes = (BytesReference) output;
+        assertThat(bytes.toUtf8(), equalTo("foo bar"));
+    }
+
+    @Test @Repeat(iterations = 5)
```
Comment (MaineC) [inline src/test/java/org/elasticsearch/script/mustache/MustacheScriptEngineTest.java]: See above

File: src/main/java/org/elasticsearch/script/mustache/InternalMustacheFactory.java [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+            case '"':
+            case '\\':
+            case '\u000B': // vertical tab
+            case '\t':
+                return true;
+        }
+        return false;
+    }
+
+    static class InternalObjectHandler extends ReflectionObjectHandler {
```
Comment (MaineC) [inline src/main/java/org/elasticsearch/script/mustache/InternalMustacheFactory.java]: Some documentation would be helpful here.

File: src/main/java/org/elasticsearch/script/mustache/InternalMustacheFactory.java [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+                } else if (object instanceof Collection) {
+                    return new CollectionMap((Collection) object);
+                }
+            }
+            return super.coerce(object);
+        }
+
+    }
+
+    static class ArrayMap extends AbstractMap<Object, Object> implements Iterable<Object> {
```
Comment (MaineC) [inline src/main/java/org/elasticsearch/script/mustache/InternalMustacheFactory.java]: Documentation missing.

File: src/main/java/org/elasticsearch/script/mustache/InternalMustacheFactory.java [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+        private final Collection col;
+
+        public CollectionMap(Collection col) {
+            this.col = col;
+        }
+
+        @Override
+        public Object get(Object key) {
+            if (key instanceof Number) {
+                return Iterables.get(col, ((Number) key).intValue());
```
Comment (MaineC) [inline src/main/java/org/elasticsearch/script/mustache/InternalMustacheFactory.java]: Assume col holds a Set: The iterator of a Set makes no guarantee wrt. to iteration order, so the object retrieved above is non-deterministic.

File: src/main/java/org/elasticsearch/script/mustache/InternalMustacheFactory.java [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+            }
+
+            @Override
+            public void remove() {
+                throw new UnsupportedOperationException("array iterator does not support removing elements");
+            }
+        }
+    }
+
+    static class CollectionMap extends AbstractMap<Object, Object> implements Iterable<Object> {
```
Comment (MaineC) [inline src/main/java/org/elasticsearch/script/mustache/InternalMustacheFactory.java]: Documentation missing.

File: src/test/java/org/elasticsearch/script/mustache/MustacheScriptEngineTest.java [hunk 10]
```diff
@@ -98,14 +103,60 @@ public void testEscapeJson() throws IOException {
...
         }
     }
 
+    @Test @Repeat(iterations = 5)
+    public void testArrayAccess() throws Exception {
+        Map<String, Object> vars = new HashMap<>();
+        Object data = randomFrom(
+                new String[] { "foo", "bar" },
+                ImmutableList.of("foo", "bar"),
+                ImmutableSet.of("foo", "bar"));
```
Comment (MaineC) [inline src/test/java/org/elasticsearch/script/mustache/MustacheScriptEngineTest.java]: I understand how retrieving the n-th object of an array or a List works - as Sets don't define an order referring to the n-th element makes no sense, does it?

File: src/main/java/org/elasticsearch/script/mustache/InternalMustacheFactory.java (hunk 1 above)
Comment (uboness) [inline src/main/java/org/elasticsearch/script/mustache/InternalMustacheFactory.java]: never looked into it... but I think that'll belong in a different PR... don't want to potentially break the escape functionality for the feature added here

File: src/test/java/org/elasticsearch/script/mustache/MustacheScriptEngineTest.java (hunk 2 above)
Comment (uboness) [inline src/test/java/org/elasticsearch/script/mustache/MustacheScriptEngineTest.java]: I can try to think of something, but not sure what value it adds really. I added this functionality for the mustache support, such that the deafult mustache engine can be used in other places (plugins) with this extended functionality... will see what I come up with

File: src/test/java/org/elasticsearch/script/mustache/MustacheScriptEngineTest.java (hunk 3 above)
Comment (uboness) [inline src/test/java/org/elasticsearch/script/mustache/MustacheScriptEngineTest.java]: sure

File: src/test/java/org/elasticsearch/script/mustache/MustacheScriptEngineTest.java (hunk 4 above)
Comment (uboness) [inline src/test/java/org/elasticsearch/script/mustache/MustacheScriptEngineTest.java]: sure

File: src/test/java/org/elasticsearch/script/mustache/MustacheScriptEngineTest.java (hunk 5 above)
Comment (uboness) [inline src/test/java/org/elasticsearch/script/mustache/MustacheScriptEngineTest.java]: sure

File: src/main/java/org/elasticsearch/script/mustache/InternalMustacheFactory.java (hunk 6 above)
Comment (uboness) [inline src/main/java/org/elasticsearch/script/mustache/InternalMustacheFactory.java]: sure

File: src/main/java/org/elasticsearch/script/mustache/InternalMustacheFactory.java (hunk 7 above)
Comment (uboness) [inline src/main/java/org/elasticsearch/script/mustache/InternalMustacheFactory.java]: sure

File: src/main/java/org/elasticsearch/script/mustache/InternalMustacheFactory.java (hunk 8 above)
Comment (uboness) [inline src/main/java/org/elasticsearch/script/mustache/InternalMustacheFactory.java]: indeed. for unsorted sets, it's non-deterministic, but it's also the user choice to use it (for sorted sets it will be deterministic). In practice, the model we create out of xcontent never holds a set, it always translates an array to a list, so I don' think we'll ever encounter a set. I added generic collections here so we'll always fall back on a behaviour (even if for unsorted sets it's non-deterministic). that said, I can change this to only support arrays, lists and sorted sets...

**Comment by uboness** [comment]
updated PR

- skipped the search template example as I don't find much value in providing it. It's less about search template support, more about adding the support in the language.
- limited this functionality to arrays, lists and sorted sets - collections where the order is deterministic.
- added javadocs

File: src/main/java/org/elasticsearch/script/mustache/InternalMustacheFactory.java (hunk 1 above)
Comment (MaineC) [inline src/main/java/org/elasticsearch/script/mustache/InternalMustacheFactory.java]: Makes sense.

File: src/main/java/org/elasticsearch/script/mustache/InternalMustacheFactory.java (hunk 8 above)
Comment (MaineC) [inline src/main/java/org/elasticsearch/script/mustache/InternalMustacheFactory.java]: > I added generic collections here so we'll always fall back on a behaviour (even if for unsorted sets it's non-
> deterministic). that said, I can change this to only support arrays, lists and sorted sets...

For context: Despite having had trouble understanding the semantics here IMHO this is one of the cases where failing fast is best. Probably I've encountered too many instances where instead of reading the docs people just "did whatever worked on first try" only to run into issues later.

Comment (uboness) [inline src/main/java/org/elasticsearch/script/mustache/InternalMustacheFactory.java]: so I think here failing fast has a different meaning. In the sense that it seems to be that the default behaviour of `mustache` is to ignore unrecognized keys and then return an empty string. I believe we should be consistent here and not fail fast, but rather "fallback fast" :)... fallback on the default behaviour of `mustache`. This is what I do right now (after the update)... if it's non-deterministic, I fall back on the default handling of mustache.

File: src/test/java/org/elasticsearch/script/mustache/MustacheScriptEngineTest.java (hunk 2 above)
Comment (MaineC) [inline src/test/java/org/elasticsearch/script/mustache/MustacheScriptEngineTest.java]: > such that the deafult mustache engine can be used in other places (plugins) with this 
> extended functionality.

Just out of curiosity: Do you have a specific plugin in mind that already has this need? Not sure how we go about documentation for plugin authors - if mustache engine re-use is one of the major goals here IMHO there should be some doc/example for it? 

I guess I've seen too much unused code lingering in projects in the past that turned out to be "not quite what we need" which is why I try so hard to find real use cases for verification.

File: src/main/java/org/elasticsearch/script/mustache/InternalMustacheFactory.java (hunk 8 above)
Comment (s1monw) [inline src/main/java/org/elasticsearch/script/mustache/InternalMustacheFactory.java]: can we maybe just support Lists and Arrays here I'd feel much better about it and we could remove implementations? I also wonder if we just anna turn all collections into an array instead and only have one impl?

**Comment by s1monw** [comment]
I think this looks good to me but I'd really like to get rid of the `CollectionMap` and maybe just transform the sorted collections into an array and then use the array one?

**Comment by s1monw** [comment]
@uboness can you revisit this we wanna get 1.6 out of the door soonish

**Comment by markharwood** [comment]
Nice to have added array support but perhaps missing a feature?
The ["trailing comma" issue] (http://stackoverflow.com/questions/6114435/in-mustache-templating-is-there-an-elegant-way-of-expressing-a-comma-seperated-l#comment48696688_6128029) looks to affect a lot of users and requires knowledge of what is the last element in an array. If we could have syntax for declaring the last element in an array users could include a `{{^last}}` type clause around the comma.

**Comment by clintongormley** [comment]
This PR is way out of date. Closing
